rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc

```diff
@@ -0,0 +1,171 @@
+    ws = workspaceInfo(folders, getPathConfig);
+
+    for (f <- folders) {
+        step("loading modules for project \'<f.file>\'", 1);
+        PathConfig pcfg = getPathConfig(f);
+        RascalCompilerConfig ccfg = getRascalCoreCompilerConfig(pcfg);
+
+        ms = rascalTModelForLocs(toList(find(f, "rsc")), ccfg, dummy_compile1);
+        checkNoErrors(ms);
+
+        for (m <- ms.tmodels) {
+            tm = convertTModel2PhysicalLocs(ms.tmodels[m]);
+            ws = loadModel(ws, tm);
+            ws.modules += {ms.moduleLocs[m].top};
+        }
+    }
+
+    return ws;
+}, totalWork = size(folders));
+
+@memo
+set[loc] getUses(WorkspaceInfo ws, loc def) = invert(ws.useDef)[def];
+
+@memo
+set[loc] getUses(WorkspaceInfo ws, set[loc] defs) = invert(ws.useDef)[defs];
+
+@memo
+set[loc] getDefs(WorkspaceInfo ws, loc use) = ws.useDef[use];
+
+set[loc] getOverloadedDefs(WorkspaceInfo ws, set[loc] defs, MayOverloadFun mayOverloadF) {
+    set[loc] overloadedLocs = defs;
+
+    // Pre-condition
+    assert mayOverloadF(overloadedLocs, ws.definitions):
+        "Initial defs are invalid overloads!";
+
+    for (loc d <- ws.definitions) {
+        if (mayOverloadF(defs + d, ws.definitions)) {
+            overloadedLocs += d;
+        }
+    }
+
+    return overloadedLocs;
+}
+
+tuple[set[loc], set[loc]] getDefsUses(WorkspaceInfo ws, cursor(use(), l, _), MayOverloadFun mayOverloadF) {
+    defs = getOverloadedDefs(ws, getDefs(ws, l), mayOverloadF);
+    uses = getUses(ws, defs);
+    return <defs, uses>;
+}
+
+tuple[set[loc], set[loc]] getDefsUses(WorkspaceInfo ws, cursor(def(), l, _), MayOverloadFun mayOverloadF) {
+    defs = getOverloadedDefs(ws, {l}, mayOverloadF);
+    uses = getUses(ws, defs);
+    return <defs, uses>;
+}
+
+tuple[set[loc], set[loc]] getDefsUses(WorkspaceInfo ws, cursor(typeParam(), cursorLoc, cursorName), MayOverloadFun _) {
+    AType at = ws.facts[cursorLoc];
+    if(Define d: <_, _, _, _, _, defType(afunc(_, /at, _))> <- ws.defines, isContainedIn(cursorLoc, d.defined)) {
+        // From here on, we can assume that all locations are in the same file, because we are dealing with type parameters and filtered on `isContainedIn`
+        facts = {<l, ws.facts[l]> | l <- ws.facts
+                                  , at := ws.facts[l]
+                                  , isContainedIn(l, d.defined)};
+
+        formals = {l | <l, f> <- facts, f.alabel != ""};
+
+        // Given the location/offset of `&T`, find the location/offset of `T`
+        offsets = sort({l.offset | l <- facts<0>});
+        nextOffsets = toMapUnique(zip2(prefix(offsets), tail(offsets)));
+
+        loc sentinel = |unknown:///|(0, 0, <0, 0>, <0, 0>);
+        defs = {(sentinel | (it == sentinel || f.length < it.length) && f.offset == nextOffset ? f : it | f <- facts<0>) | formal <- formals, nextOffsets[formal.offset]?, nextOffset := nextOffsets[formal.offset]};
```

Comment:
sometimes, a reducer does not make for a concise and readable implementation. I mean, nice sollution, but complex.

I would expect that if this comprehension is written as a loop with a bit of state, it might be a lot more readable. At the very least, move the `f.offset == nextOffset` into the "loop" part of the reducer.

Reply:
I rewrote this using a library function for the reduction part.